### ISSUE TITLE
ast and others not considered standard lib

### ISSUE THREAD

**Issue by Pawamoy** [timothycrosley/isort#540]
Hi Tim,

Upgrading from 4.2.5 to 4.2.8, it seems that isort now does not recognize ast as a standard library anymore.
We can see it's doc [here](https://docs.python.org/3.5/library/ast.html) in standard library so I think it really should be considered stdlib :/

What do you think?

**Comment by timgraham** [comment]
Also seeing this with ctypes, cgi, ssl, mimetypes, email, socketserver, ipaddress, and others. with Django.

**Comment by Pawamoy** [comment]
Hahaha, three Tim here. Thanks, I'll update the issue title.

**Comment by joshp123** [comment]
I'm also experiencing this issue, it seems like many modules are missing from the 'known_standard_library' key in the configuration:
```
Python 3.5.2 (default, Nov 17 2016, 17:05:23)
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from isort import isort
>>> i = isort.SortImports()
>>> i.config['known_standard_library']
['abc', 'anydbm', 'argparse', 'array', 'asynchat', 'asyncio', 'asyncore', 'atexit', 'base64', 'BaseHTTPServer', 'bisect', 'bz2', 'calendar', 'cgitb', 'cmd', 'codecs', 'collections', 'commands', 'compileall', 'concurrent', 'ConfigParser', 'contextlib', 'Cookie', 'copy', 'cPickle', 'cProfile', 'cStringIO', 'csv', 'datetime', 'dbhash', 'dbm', 'decimal', 'difflib', 'dircache', 'dis', 'doctest', 'dumbdbm', 'EasyDialogs', 'errno', 'exceptions', 'filecmp', 'fileinput', 'fnmatch', 'fractions', 'functools', 'ftplib', 'gc', 'gdbm', 'getopt', 'getpass', 'gettext', 'glob', 'grp', 'gzip', 'hashlib', 'heapq', 'hmac', 'html', 'http', 'imaplib', 'imp', 'inspect', 'io', 'itertools', 'json', 'linecache', 'locale', 'logging', 'mailbox', 'math', 'mhlib', 'mmap', 'multiprocessing', 'numbers', 'operator', 'optparse', 'os', 'pdb', 'pickle', 'pipes', 'pkgutil', 'platform', 'plistlib', 'pprint', 'profile', 'pstats', 'pwd', 'pyclbr', 'pydoc', 'Queue', 'random', 're', 'readline', 'resource', 'rlcompleter', 'robotparser', 'sched', 'select', 'shelve', 'shlex', 'shutil', 'signal', 'SimpleXMLRPCServer', 'site', 'sitecustomize', 'smtpd', 'smtplib', 'socket', 'SocketServer', 'sqlite3', 'string', 'StringIO', 'struct', 'subprocess', 'sys', 'sysconfig', 'tabnanny', 'tarfile', 'tempfile', 'textwrap', 'threading', 'time', 'timeit', 'trace', 'traceback', 'typing', 'unittest', 'urllib', 'urllib2', 'urlparse', 'usercustomize', 'uuid', 'warnings', 'weakref', 'webbrowser', 'whichdb', 'xml', 'xmlrpc', 'xmlrpclib', 'zipfile', 'zipimport', 'zlib', 'builtins', '__builtin__', 'thread', 'binascii', 'statistics', 'unicodedata', 'fcntl', 'pathlib', 'wsgiref', 'syslog']
>>> 'enum' in i.config['known_standard_library']
False
>>> 'configparser' in i.config['known_standard_library']
False
>>>
```

I suspect the changes are from the following block:
https://github.com/timothycrosley/isort/compare/4.2.5...master#diff-7597e4974079b411e5fb6464a449abf1R263 but I haven't been able to confirm.

**Comment by timgraham** [comment]
Bisected to 5d62a73b5b3edef073edf5e58c977c89ddd961f4, \cc @nicoddemus as the author of that commit.

**Comment by nicoddemus** [comment]
Does that happen in a virtual environment or a standard installation? IIRC there's a difference on the layout between the two.

**Comment by timgraham** [comment]
Virtual environment here.

**Comment by joshp123** [comment]
I'm experiencing this in a virtual environment.

**Comment by nicoddemus** [comment]
In summary my change:

1. Compares the paths taking in account the case (so `ABC` does not resolve to the `abc` module on Windows just because the path does); this should not affect Linux because it only does that for Windows;
2. Using `sysconfig.get_paths()['stdlib']` to get the path to the standard lib instead of check if `"python2"` or `"python3"` is somewhere in the path to the module.

Regarding 2:

In a virtual environment at `/home/jenkins/.tmp-env`:

```
>>> import sysconfig
>>> sysconfig.get_paths()['stdlib']
'/home/jenkins/.tmp-env/lib/python2.7'
```

`ast` does not exist in `/home/jenkins/.tmp-env/lib/python2.7`, and apparently virtualenv does some hack behind the scenes to fallback to the standard installation for some modules like `os` and `ast`:

```
>>> import os
>>> os.__file__
'/home/jenkins/.tmp-env/lib/python2.7/os.pyc'
>>> import ast
>>> ast.__file__
'/usr/lib/python2.7/ast.pyc'
```

So the previous check for `python2.7` worked, which explains the regression.

My suggestion is to just add those "special" modules to the `known_standard_library`, I guess it is there exactly to circumvent these cases.

**Comment by nicoddemus** [comment]
> My suggestion is to just add those "special" modules to the known_standard_library

In case it was not clear, I meant adding this to `isort`'s defaults, not making each user add them; 😁

**Comment by Pawamoy** [comment]
Indeed, only happening in a virtualenv. I previously tried adding ast into `known_stdlib` as I didn't know it was `known_standard_library`. It worked for me.

**Comment by nicoddemus** [comment]
Anybody would care opening a PR adding at least the modules mentioned in this thread to `known_standard_library`? 😁

**Comment by timgraham** [comment]
If possible, a more comprehensive solution or determination of modules to be added to `known_standard_library` would be nice. It doesn't seem ideal to rely on user bug reports to populate that list.

**Comment by nicoddemus** [comment]
> If possible, a more comprehensive solution or determination of modules to be added to known_standard_library would be nice

Definitely.

**Comment by timothycrosley** [comment]
Working on get out a hot fix for this:

https://github.com/timothycrosley/isort/pull/544/files

I've added all stdlib modules defined here: https://docs.python.org/3/py-modindex.html to the list. Beyond this, I did notice a problem with your change @nicoddemus, and that's that isort is never supposed to care about the version of Python it's being run in, which that change requires. The reason for this is that when you run isort on a CI system accross multiple versions of Python, you always want the same result - since you can't change imports per system.

Thanks!

~Timothy

**Comment by timgraham** [comment]
That branch fixes all instances of this problem in Django.

**Comment by nicoddemus** [comment]
> isort is never supposed to care about the version of Python it's being run in, which that change requires. The reason for this is that when you run isort on a CI system accross multiple versions of Python, you always want the same result - since you can't change imports per system.

Makes sense, but I'm not aware of an alternative for that case because it uses the non-deprecated form when possible. But I guess it is not much of a problem given that isort is supposed to drop Python 2.6 in 4.3. 👍

**Comment by timothycrosley** [comment]
I've pushed out a new hot fix release: `4.2.9` that should fix this issue by expanding the list of known stdlib. Thanks everyone!
